### PR TITLE
Use RemoveSafeDeltaCounterCell for PerWorker counters

### PR DESCRIPTION
**Pull request by JayajP**
Dataflow Streaming harness's metrics container will store `per worker counters` in a new `RemoveSafeDeltaCounterCell`.

For the new `PerWorkerMetrics` we want to support dynamically creating metrics that may not live the lifetime of the job, therefore we need to implement garbage collection for stale metrics.

E.g. We want to support BigQuery Metrics on a per-bigquery-table basis. If the pipeline uses dynamic destinations to write to 100 tables at 12:00 and at 15:00 is only writing to a few tables, the underlying MetricsContainer should not hold on to metrics for all 100 tables.

In`RemoveSafeDeltaCounterCell` metric updates and deletions are handled atomically - using `ConcurrentHashMap::compute` and `ConcurrentHashMap::remove` respectively - ensuring safe garbage collection.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

**Comment by AnandInguva** [comment]
assign to next reviewer

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/RemoveSafeDeltaCounterCell.java [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.runners.dataflow.worker;
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicLong;
+import javax.annotation.Nonnull;
```
Comment (Abacn) [inline runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/RemoveSafeDeltaCounterCell.java:22]: Nonnull is default and should not be necessary. Also we use checkframework's Nonnull, not Java one

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/RemoveSafeDeltaCounterCell.java [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+/**
+ * Version of {@link CounterCell} supporting multi-thread safe mutations and extraction of delta
+ * values. {@link RemoveSafeDeltaCounterCell} allows safe deletions from the underlying {@code
+ * countersMap} through the {@code deleteIfZero} method.
+ */
+public class RemoveSafeDeltaCounterCell implements Counter, MetricCell<Long> {
+
+  private final MetricName metricName;
+  private final ConcurrentHashMap<MetricName, AtomicLong> countersMap;
```
Comment (Abacn) [inline runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/RemoveSafeDeltaCounterCell.java:37]: It is important to note that this class does not own countersMap. Its more like a wrapper operating a single key in countersMap with Counter semantics.

This looks a little bit unusual to me at first glance. May consider add some comments here

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/RemoveSafeDeltaCounterCell.java [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+  @Override
+  public Long getCumulative() {
+    throw new UnsupportedOperationException("getCumulative is not supported by Streaming Metrics");
+  }
+
+  /** Remove the metric from the {@code countersMap} if the the metric is zero valued. */
+  @SuppressWarnings(
+      "nullness") // computeIfPresent's remapping function can return null to remove the entry.
+  public void deleteIfZero() {
```
Comment (Abacn) [inline runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/RemoveSafeDeltaCounterCell.java:102]: Seems this is not yet used in other main scope codes. Is it planned as the next step?

**Review by Abacn — COMMENTED** [review]
Thanks, left a few comments

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/RemoveSafeDeltaCounterCell.java (hunk 2 above)
Comment (JayajP) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/RemoveSafeDeltaCounterCell.java:41]: Done

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/RemoveSafeDeltaCounterCell.java (hunk 1 above)
Comment (JayajP) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/RemoveSafeDeltaCounterCell.java]: Done

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/RemoveSafeDeltaCounterCell.java (hunk 3 above)
Comment (JayajP) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/RemoveSafeDeltaCounterCell.java:110]: Yes this is next step, need to wait for my v1b3 api changes to be rolled out to the public api.

**Comment by Abacn** [comment]
Java PreCommit failing test due to https://github.com/apache/beam/issues/28957#issuecomment-1871410640

**Review by Abacn — APPROVED** [review]

**Pull request merged**